kotlin/viewshed-geoprocessing/src/main/java/com/esri/arcgisruntime/sample/viewshedgeoprocessing/MainActivity.kt

```diff
@@ -0,0 +1,216 @@
+        addGraphicForPoint(mapPoint)
+        calculateViewshedAt(mapPoint)
+        return super.onSingleTapConfirmed(e)
+      }
+    }
+  }
+
+  /**
+   * Adds a graphic at the chosen mapPoint.
+   *
+   * @param point in MapView coordinates.
+   */
+  private fun addGraphicForPoint(point: Point) {
+    // remove existing graphics
+    inputGraphicsOverlay.graphics.clear()
+
+    // add new graphic to the graphics overlay
+    inputGraphicsOverlay.graphics.add(Graphic(point))
+  }
+
+  /**
+   * Uses the given point to create a FeatureCollectionTable which is passed to performGeoprocessing.
+   *
+   * @param point in MapView coordinates.
+   */
+  private fun calculateViewshedAt(point: Point) {
+    // remove previous graphics
+    resultGraphicsOverlay.graphics.clear()
+
+    // cancel any previous job
+    geoprocessingJob?.cancel()
+
+    // create field with same alias as name
+    val field = Field.createString("observer", "", 8)
+
+    // create feature collection table for point geometry
+    val featureCollectionTable =
+      FeatureCollectionTable(listOf(field), GeometryType.POINT, point.spatialReference)
+    featureCollectionTable.loadAsync()
+
+    // create a new feature and assign the geometry
+    val newFeature = featureCollectionTable.createFeature().apply {
+      geometry = point
+    }
+
+    // add newFeature and call perform Geoprocessing on done loading
+    featureCollectionTable.addFeatureAsync(newFeature)
+    featureCollectionTable.addDoneLoadingListener {
+      if (featureCollectionTable.loadStatus == LoadStatus.LOADED) {
+        performGeoprocessing(featureCollectionTable)
+      }
+    }
+  }
+
+  /**
+   * Creates a GeoprocessingJob from the GeoprocessingTask. Displays the resulting viewshed on the map.
+   *
+   * @param featureCollectionTable containing the observation point.
+   */
+  private fun performGeoprocessing(featureCollectionTable: FeatureCollectionTable) {
+
+    // geoprocessing parameters
+    val parameterFuture: ListenableFuture<GeoprocessingParameters> =
+      geoprocessingTask.createDefaultParametersAsync()
+    parameterFuture.addDoneListener {
+      try {
+        val parameters = parameterFuture.get().apply {
+          processSpatialReference = featureCollectionTable.spatialReference
+          outputSpatialReference = featureCollectionTable.spatialReference
+
+          // use the feature collection table to create the required GeoprocessingFeatures input
+          inputs["Input_Observation_Point"] = GeoprocessingFeatures(featureCollectionTable)
+        }
+
+        // initialize job from geoprocessingTask
+        geoprocessingJob = geoprocessingTask.createJob(parameters)
+
+        // start the job
+        geoprocessingJob?.start()
+
+        // listen for job success
+        geoprocessingJob?.addJobDoneListener {
+          if (geoprocessingJob?.status == Job.Status.SUCCEEDED) {
+            val geoprocessingResult = geoprocessingJob?.result
```

Comment:
Could possibly inline this:

```
// get the viewshed from geoprocessingResult
            (geoprocessingJob?.result?.outputs?.get("Viewshed_Result") as? GeoprocessingFeatures)?.let { viewshedResult ->
              // for each feature in the result
              for (feature in viewshedResult.features) {
                // add the feature as a graphic
                resultGraphicsOverlay.graphics.add(Graphic(feature.geometry))
              }
            }
```

Reply:
I'm struggling to imagine coming up with this on my own!